### DISCUSSION TITLE
Connecting to CodeSpace with mosh

### DISCUSSION THREAD

**Discussion by bdubaut** [community/community#26081] [Codespaces] [answered]
<p>Hi,</p>
<p>I own an iPad and there is times where I need to write code. How can I connect to the codespace and use it with <a href="https://mosh.org/" rel="nofollow noopener">Mosh</a>? I tried to use the codespaces interface in the chrome mobile browser but it’s working super-well (aspect ratio, keyboard shortcuts, etc.).</p>
<p>Thanks!</p>

**Comment by nethgato** [comment, marked as answer]
<p>Hiii <a class="mention" href="/bdubaut">@bdubaut</a> thanks for posting! I had to look into what Mosh was, but that seems so rad! TIL =)</p>
<p>Unfortunately though, I don’t think we’d be able to get this to happen. Check:</p>
<ul>
<li><a href="https://github.com/microsoft/vscode-remote-release/issues/334">This issue</a></li>
</ul>
<p>…on the VSCode repo.</p>
<blockquote>
<p>vscode remote ssh requires ssh port forwarding to access the agent. AFAIK mosh doesn’t provide such a feature.</p>
</blockquote>
<p>Granted, it’s been some time, this does seem to:</p>
<ul>
<li><a href="https://github.com/mobile-shell/mosh/issues/337">still be accurate</a></li>
</ul>
<p>Dang =(</p>

**Comment by bdubaut** [comment]
<p><a class="mention" href="/nethgato">@nethgato</a>  hmm that’s a shame! Thanks for looking it up!</p>

**Comment by dtinth** [comment]
I am able to use Codespaces with `mosh` now. All I need to do is to put both computers on the same VPN, so they can talk to each other without either needing a public IP. But you will need a VPN server with a public IP to act as a communication hub.

I'll share my setup:

**Prerequisite:**

- A VPS to act as a VPN
    - a publicly accessible IP
    - Docker installed
    - [wg-easy](https://github.com/wg-easy/wg-easy) set up
- gh cli

**Setup:**

1. Use `gh codespace ssh` to connect to the codespace through normal means.
2. Install Wireguard and Mosh in codespace: `sudo apt install wireguard mosh`
3. Create two configurations - one for your machine, another for the codespace.
4. Modify the configuration per below section
5. Use Wireguard app to load the configuration and connect your computer to the VPN
6. Put the codespace's configuration file into `/etc/wireguard/wg0.conf`
7. Connect the codespace to the VPN with `sudo wg-quick up wg0`

After step 7, I can now ping my Codespace through its VPN IP address. I can then use mosh like this:

```
mosh --ssh="ssh -p 2222" vscode@10.8.0.3
```

Note that step 7 has to be repeated each time the Codespace boots up. Another upside is that now you can also connect to any service running on the Codespace without having to forward ports manually.

**Config file modification**

Before:

```
[Interface]
PrivateKey = kG0G9DV0CdbCEPf6qW5/PxUpI62V/VEMbb27ogXa5Gk=
Address = 10.8.0.4/24
DNS = 1.1.1.1

[Peer]
PublicKey = XRandOMusJWGjUpJBOFn8KHu8/RANDOMoKB+jH20GG0=
PresharedKey = RanDom3+RanD+RanDoMxvWPhGQvjBidZUF/XbjMJAhg=
AllowedIPs = 0.0.0.0/0, ::/0
PersistentKeepalive = 0
Endpoint = ___.___.___.___:51820
```

I make these changes:

- Remove `DNS`
- Change `AllowedIPs` to `10.8.0.0/24`
- Change `PersistentKeepalive = 25`

After:

```
[Interface]
PrivateKey = kG0G9DV0CdbCEPf6qW5/PxUpI62V/VEMbb27ogXa5Gk=
Address = 10.8.0.4/24

[Peer]
PublicKey = XRandOMusJWGjUpJBOFn8KHu8/RANDOMoKB+jH20GG0=
PresharedKey = RanDom3+RanD+RanDoMxvWPhGQvjBidZUF/XbjMJAhg=
AllowedIPs = 10.8.0.0/24
PersistentKeepalive = 25
Endpoint = ___.___.___.___:51820
```
